파이썬에서 함수는 크게 내장 함수(Built-in Functions)와 외장 함수(External Functions)로 나뉩니다.

1. 내장 함수 (Built-in Functions):
- 내장 함수는 파이썬 인터프리터에 포함되어 있어, 별도의 모듈을 임포트(import)하지 않고도 사용 예를 들어, print(), len(), type(), range(), zip() 등이 내장 함수입니다.
- 이러한 함수들은 기본적인 작업을 수행하며, 파이썬 프로그래밍의 기본적인 빌딩 블록(building blocks)으로 작동

2. 외장 함수 (External Functions):
- 표준 라이브러리에 포함된 모듈과 함수를 가리키는 것으로 모듈을 임포트하여 사용
- 파이썬 표준 라이브러리의 os, sys, math, random, datetime 등의 모듈에는 다양한 외장 함수가 포함

3. 외부 라이브러리나 패키지
- 표준 라이브러리에 포함되지 않아 기본 설치에는 포함되지 않는다. 
- 이를 사용하려면 먼저 해당 라이브러리나 패키지를 설치
- 외부 라이브러리는 통상적으로 파이썬의 패키지 관리자인 pip를 사용하여 설치
- 설치 후에는 외장 함수와 마찬가지로 필요한 모듈을 임포트하여 사용
- 예: numpy, pandas, requests 등의 외부 라이브러리에 포함된 함수들.

3. 사용자 정의 함수(User-Defined Functions):
- 사용자가 직접 정의하여 사용하는 함수
- def 키워드를 사용하여 함수를 정의하며, 필요에 따라 매개변수와 반환 값을 지정

4. 람다 함수(Lambda Functions):
- 이름 없는 간단한 함수를 정의할 때 사용하는 익명 함수
- lambda 키워드를 사용하여 한 줄로 함수를 정의하며, 주로 간단한 연산이 필요할 때 사용

In [2]:
# 내장 함수 예제
print("Hello, World!")  # 출력 함수
print(len([1, 2, 3, 4]))  # 길이 계산 함수
print(type(10))  # 타입 확인 함수


Hello, World!
4
<class 'int'>


In [23]:
# 외장 함수 예제
import math  # 수학 모듈 임포트
print(math.sqrt(16))  # 제곱근 계산 함수

import random  # 랜덤 모듈 임포트
print(random.randint(1, 10))  # 랜덤 정수 생성 함수. 양끝 포함

import datetime  # 날짜와 시간 모듈 임포트
print(datetime.datetime.now())  # 현재 날짜와 시간 출력 함수


4.0
3
2023-10-31 15:13:04.537211


In [26]:
# 외부 라이브러리 예제
# 먼저 pip를 사용하여 requests 라이브러리를 설치합니다.
# !pip install pandas

import pandas as pd

# pandas의 DataFrame 객체 생성
data = {'Name': ['Alice', 'Bob', 'Charlie'], 'Age': [25, 30, 35]}
df = pd.DataFrame(data)

# pandas의 head() 함수 사용하여 DataFrame의 처음 5행 출력 (여기서는 3행만 존재)
print(df.head(2))


    Name  Age
0  Alice   25
1    Bob   30


In [28]:
# 사용자 함수 예제
def greet(name):
    return f"Hello, {name}!"

print(greet("한정현"))  # 출력: Hello, Alice!


Hello, 한정현!


#### 람다함수

In [29]:
# 람다 함수는 간단한 한 줄의 함수를 작성할 때 유용
multiply = lambda x, y: x * y
print(multiply(3, 4))  # 출력: 12


12


In [30]:
# 조건 사용:
is_even = lambda x: True if x % 2 == 0 else False
print(is_even(4))  # 출력: True
print(is_even(5))  # 출력: False


True
False


In [35]:
# map() 함수와 함께 사용:
numbers = [1, 2, 3, 4, 5]
squared_numbers = map(lambda x: x * x, numbers)
print(list(squared_numbers))  # 출력: [1, 4, 9, 16, 25]


[1, 4, 9, 16, 25]


In [37]:
# filter() 함수와 함께 사용:
numbers = [1, 2, 3, 4, 5]
even_numbers = filter(lambda x: x % 2 == 0, numbers)
print(list(even_numbers))  # 출력: [2, 4]


[2, 4]


In [4]:
# sorted() 함수와 함께 사용:
students = [('Alice', 90), ('Bob', 85), ('Charlie', 88)]
# sorted_students = sorted(students, key=lambda x: x[1]) # key는 sort 기준
sorted_students = sorted(students, key=lambda x: x[1], reverse=True) # 내림차순
print(sorted_students)

[('Alice', 90), ('Charlie', 88), ('Bob', 85)]


과제_8_1031
- 내장함수(배운 것) 5개를 활용해서 간단한 프로그램을 만드세요.
- 외장함수(배운 것) 5개를 활용해서 간단한 프로그램을 만드세요
- 사용자 함수 3개를 작성하세요
- 리스트를 작성하고 람다 함수를 이용해서 한번에 함수를 적용하여 결과값을 출력하세요
- 외부 라리브러리를 임포트해서 간단한 프로그램을 만드세요.


In [6]:
import pandas as pd
import seaborn as sns

titanic = sns.load_dataset('titanic')

titanic.head()


survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

#### 함수 vs 메소드 
- 함수 (Function)
    - 함수는 특정 작업을 수행하거나 값을 계산하기 위해 정의된 코드의 블록
    - 함수는 def 키워드를 사용하여 정의
    - 함수는 전역적으로 호출할 수 있거나 다른 함수 내부에서 호출할 수 있다.
    - 예: print(), len(), max()
- 메소드 (Method)
    - 메소드는 객체에 속한 함수즉, 메소드는 특정 객체의 속성이거나 기능으로 정의
    - 메소드는 객체와 연결되어 있으며, . (점) 표기법을 사용하여 해당 객체에 대한 메소드를 호출
    - 메소드는 첫 번째 매개변수로 self를 사용하여 객체 자체를 참조 (클래스 내부에서 메소드를 정의할 때)
- 주요 차이점
    - 정의: 함수는 독립적으로 정의되지만, 메소드는 클래스 내부에서 정의
    - 호출: 함수는 이름만으로 호출될 수 있지만, 메소드는 객체를 통해 호출
    - 참조: 메소드는 첫 번째 매개변수로 객체 자신을 참조(self), 반면 함수는 이런 자체 참조 매개변수를 갖지 않는다.

In [6]:
def my_function():
    print("Hello from a function")

my_function()  # 함수 호출

Hello from a function


In [7]:
class MyClass:
    def my_method(self):
        print("Hello from a method")

obj = MyClass()
obj.my_method()  # 메소드 호출

Hello from a method


In [11]:
# 메소드 사례
my_list = [1, 2, 3, 4, 5]

my_list.append(6)     # 리스트에 요소 추가
my_list

[1, 2, 3, 4, 5, 6]

In [8]:
my_string = "hello"

upper_string = my_string.upper()  # 문자열을 대문자로 변환
upper_string

'HELLO'

In [7]:
my_dict = {'key1': 'value1', 'key2': 'value2'}

value = my_dict.get('key1')  # key1에 해당하는 값을 가져옴
value

'value1'

In [11]:
my_set = {1, 2, 3}

my_set.add(4)  # 집합에 요소 추가
my_set

{1, 2, 3, 4}

In [9]:
my_list = [1, 2, 3, 4, 5]

index = my_list.index(3)  # 리스트에서 3이라는 값의 인덱스 반환
index

2

#### 사용자 함수

In [12]:
def add_number(n1, n2):
   ret = n1+n2
   return ret

def add_txt(t1, t2):
   print(t1+t2)
 
ans1 = add_number(10, 15)
print(ans1)                   # 25가 출력됨
ans2 = add_number(20, 25)
print(ans2)   
text1 = '대한민국~'
text2 ='만세!!'
add_txt(text1, text2)         # ‘대한민국~만세!!’가 출력됨

25
45
대한민국~만세!!


In [16]:
# 함수 인자 : 기본적으로 인자 순서에 대응되는 값을 함수 코드에 대입
def add_txt(t1, t2='파이썬'): # t2 기본값을 대입한 기본 인자는 마지막에 위치해야 함
   print(t1+' : ' +t2)
   
add_txt('베스트')                   # ‘베스트 : 파이썬’이 출력됨. t1은 베스트, t2는 기본값이 대입
add_txt(t2='대한민국', t1='1등')    # ‘1등 : 대한민국’이 출력됨. 키워드 인자에 의한 값 전달. 인자의 순서는 무시

def func1(*args): # 인자의 개수가 불명확한 경우 가변 인자 사용. args는 함수 내부에서 튜플로 처리
   print(args)

def func2(width, height, **kwargs): # 키워드 인자가 불명확한 경우 **kwargs를 사용. 함수 내부에서 사전으로 처리
   print(kwargs)

func1()                          # 빈 튜플 ( ) 이 출력됨
func1(3,5,7,8,9)                  # (3, 5, 1, 5)가 출력됨
func2(10, 20)                          # 빈 사전 { }이 출력됨
func2(10, 20, depth=50, color='blue')  #{‘depth’:50, ‘color’:’blue’} 이 출력됨

베스트 : 파이썬
1등 : 대한민국
()
(3, 5, 7, 8, 9)
{}
{'depth': 50, 'color': 'blue'}


In [5]:
def func2(width, height, **kwargs):
    return width, height, kwargs

# func2(10, 20) 
func2(10, 20, depth=50, color='blue') 

(10, 20, {'depth': 50, 'color': 'blue'})

In [18]:
# Q. 1~10까지 가변적으로 연산을 할 수 있는 사용자 함수를 작성(매개변수가 가변적)
def all_add(*args):
    result = 0
    for i in args:
        result += i
    print(result)

all_add(4,22,59)
all_add(4,20,50,65)
all_add(4,20,50,65,75)

85
139
214


In [21]:
# Q. **args를 이용하여 아래와 같이 출력할 수 있는 사용자 함수 func2를 작성하세요
# 출력:
# Width: 100
# Height: 200
# Other arguments: {'color': 'red', 'thickness': 5}

def func2 (width, height, **kwargs):
    print(f'Width: {width}')
    print(f'Height: {height}')
    print(f'Other arguments: {kwargs}')

func2(100,200,color='red', thickness=5)
func2(100,200,color='red', thickness=5, type='large')

Width: 100
Height: 200
Other arguments: {'color': 'red', 'thickness': 5}
Width: 100
Height: 200
Other arguments: {'color': 'red', 'thickness': 5, 'type': 'large'}


In [ ]:
# 과제1_1101 : 사용자 함수를 아래와 같이 사용할 수 있는 func3를 작성하세요.
func3(100, 200)  
# 출력: Width: 100, Height: 200, Color: blue, Thickness: 1

func3(100, 200, color='green', thickness=10)  
# 출력: Width: 100, Height: 200, Color: green, Thickness: 10

In [5]:
def func3(a, b,  color='blue', thickness=1):
    print(f'Width: {a}, Height: {b}, Color: {color}, Thickness: {thickness}')
    
# 출력 : Width: 100, Height: 200, Color: blue, Thickness: 1
func3(100, 200)

# 출력: Width: 100, Height: 200, Color: green, Thickness: 10
func3(100, 200, color='red', thickness=15 )

Width: 100, Height: 200, Color: blue, Thickness: 1
Width: 100, Height: 200, Color: red, Thickness: 15


In [7]:
# kwargs 연습
def func3(width, height, **kwargs):
    dic_text = str(kwargs)[1:-1].replace("'", "")
    text3=f'Width: {width}, Height: {height}, {dic_text}'
    print(text3)

func3(100,200, color='green', thickness=10)

Width: 100, Height: 200, 
Width: 100, Height: 200, color: green, thickness: 10


In [3]:
def func3(width, height, **kwargs):
    print("Width:", width)
    print("Height:", height)
    print("Other arguments:", kwargs)
func3(100, 200, color='blue', thickness=1)
func3(100, 200, color='green', thickness=10) 

Width: 100
Height: 200
Other arguments: {'color': 'blue', 'thickness': 1}
Width: 100
Height: 200
Other arguments: {'color': 'green', 'thickness': 10}


In [ ]:
# 과제2_1101 : 아래와 같이 출력되도록 func4를 작성하세요
def print_info(**kwargs):
    for key, value in kwargs.items():
        print(f"{key}: {value}")

        
func4(100, 200, color='yellow', style='dotted')
# 출력:
# color: yellow
# style: dotted

In [8]:
def func4(a,b,**kwargs):
    print_info(**kwargs)

def print_info(**kwargs):
    for key, value in kwargs.items():
        print(f"{key}: {value}")
        
func4(100, 200, color='yellow', style='dotted')

color: yellow
style: dotted


#### 인자와 리턴값 여부에 따른 함수 유형

In [22]:
# 입력값이 없고 결과값만 있는 함수
def say():
    return 'HI'

a = say()
print(a)

HI


In [25]:
# 입력값은 있고 결과값이 없는 함수
# 결과값은 오직 return 명령어로만 돌려 받을 수 있음
def add(a,b):
    print('%d, %d의 합은 %d입니다.' %(a,b,a+b))
# add(3,4)
a = add(3,4)
print(a)

3, 4의 합은 7입니다.
None


In [27]:
# 입력값도 결과값도 없는 함수

def say():
    print('Hi')
    
a = say()
print(a)

Hi
None


In [26]:
# 입력값, 결과값이 모두 있는 함수
def add(a,b):
    return a+b
a = add(3,4)
print(a)
# add(3,4)

7


In [29]:
# 입력값, 결과값이 모두 있는 함수
def max(a,b,c):
    if a>b:
        if a>c:
            return a
        elif a<c:
            return c
    if a<b:
        if b>c:
            return b
        else:
            return c

print("가장 큰 수는 %d입니다." %(max(20,1,30)))
max(10,20,30)


가장 큰 수는 30입니다.


30

In [32]:
def math(x, y, z):
    if z == '*':
        return x*y
    elif z== '/':
        return x/y
    elif z== '+':
        return x+y
    elif z== '%':
        return x%y
    elif z== '^':
        return x**y
    elif z== '-':
        return x-y
    elif z== '//':
        return x//y
    else:
        return

print(math(int(input("숫자1 : ")), int(input("숫자2 : ")), input("기호 : ")))
# math(10,2,'//')


숫자1 : 5
숫자2 : 5
기호 : +
10


In [37]:
def com(a,b):
    if a>b:
        return f'{a} bigger than {b}'
    elif a<b:
        return f'{a} smaller than {b}'
    else:
        return f'{a} equal {b}'
com(3,3)
com(3,2)

3 equal 3


'3 bigger than 2'

In [ ]:
# 함수의 반환 값은 언제나 하나임 - 결과 값으로 튜플 값 하나를 갖게 됨
def add_and_mul(a,b):
    return a+b, a*b

result = add_and_mul(3,4)
print(result)

(7, 12)


In [ ]:
# add_mul() 함수를 정의하여 1,2,3,4,5를 입력 값으로 적용하여 덧셈(add),곱셈(mul) 연산을 수행하세요.
def add_mul(choice, *args):
    
    
add_mul('add', 1,2,3,4,5)
add_mul('mul', 1,2,3,4,5)

In [38]:
def add_mul(choice, *args):
    if choice == '+':
        result = 0
        for i in args:
            result += i
        return result
    elif choice == '*':
        result = 1
        for i in args:
            result *= i
        return result
    else :
        print('수행할 수 없는 연산입니다.')

print(add_mul('*',1,2,3,4,5))
add_mul('+',1,2,3,4,5)

120


15

In [ ]:
# 과제3_1101 : 첫번째 숫자에서 뻴셈, 나눗셈 멀티 계산 함수 sub_div를 작성
('-', 2,4,6,8,10)
('/', 2,4,6,8,10)

In [13]:
def sub_div(choice, *args):
    if choice == '-':
        result = args[0]
        for n in args[1:] :
            result -= n
        return result
    elif choice == '/':
        result = args[0]
        for i in args[1:]:
            result /=i
        return result
    else:
        return "계산할 수 있는 기호가 아닙니다."
            
print(sub_div('-', 2,4,6,8,10))
print(sub_div('/', 2,4,6,8,10))
print(sub_div('/', 2,4,6,8,10,11,12))

-26
0.0010416666666666667
7.891414141414141e-06


In [12]:
def sub_div (choice, *args) :
    if choice == "sub":
        result = args[0]
        for i in args[1:] : 
            result -= i
        return result
    elif choice == 'div':
        result = args[0]
        for i in args[1:] :
            if i == 0:
                return print("오류")
            result /= i
        return round(result,4) 

print(sub_div('sub', 2,4,6,8,10))
print(sub_div('div', 2,4,6,8,10))


-26
0.001


#### 지역변수와 전역변수
- 지역변수는 함수 내부에서만 유효하고 
- 전역변수는 코드 전반에 걸쳐 유효
- 함수의 인자는 지역변수

In [39]:
strdata = '전역변수'

def func1():
   strdata = '지역변수'
   print(strdata)
    
func1()                  # ‘지역변수’가 출력됨

지역변수


In [40]:
print(strdata)           # ‘전역변수’가 출력됨

전역변수


In [22]:
param = 10
def func2(param):       
   param = 1

print(param)             # 전역변수 10이 출력됨

10


In [42]:
param = 10
def func3():
   global param   # 함수 내부에서 전역변수를 사용하려면 'global' 키워드를 이용해 전역변수를 사용한다고 명시
   param = 50
func3()
print(param)      # 50이 출력됨

50


In [45]:
param = 10
strdata = '전역변수'

def func1():
   strdata = '지역변수'
   print(strdata)
   
def func2(param):
   param = 1
   print(param)
def func3():
   global param
   param = 50
   
func1()                  # ‘지역변수’가 출력됨
print(strdata)           # ‘전역변수’가 출력됨
print(param)             # 10이 출력됨
func2(param)
print('전역변수를 함수안에서 사용')
func3()
print(param)             # 50이 출력됨

지역변수
전역변수
10
1
전역변수를 함수안에서 사용
50


In [14]:
x =10

def modify_global():
    global x    
    x = 20
    print("함수 내에서의 x 값:", x)

modify_global()
print("함수 밖에서의 x 값:", x)

함수 내에서의 x 값: 20
함수 밖에서의 x 값: 20


In [ ]:
# 과제4_1101 : 아래 사용자 함수에서 결과가 아래와 같이 출력되도록 완성하세요.
함수 내에서의 x 값: 20
함수 밖에서의 x 값: 20

In [46]:
x = 10

def modify_global():
    x = 20
    print("함수 내에서의 x 값:", x)
modify_global()
print("함수 밖에서의 x 값:", x)

함수 내에서의 x 값: 20
함수 밖에서의 x 값: 10


In [ ]:
# 함수 안에서 선언된 변수의 효력 범위
a = 1
def vartest(a):    
    a = a + 1
    
print(vartest(a))
print(a)

None
1


In [ ]:
# 함수 안에서 함수 밖의 변수를 변경하는 방법 - return 이용

a = 1
def vartest(a):
    a =  a+1
    return a
a = vartest(a)
print(a)

2


In [47]:
# global 명령을 이용하는 방법
a = 1
def vartest():
    global a
    a = a + 1
    
vartest()
print(a)

2


#### 함수 리턴값 : return
- 함수는 한 개 이상의 값을 리턴할 수 있으며 리턴값이 없을 수도 있다.
- 리턴 값이 여러 개인 경우에는 튜플로 리턴값을 만들어 리턴

In [26]:
def reverse(x, y, z):
   return z, y, x

ret = reverse(1, 2, 3)
print(ret)                       # (3, 2, 1)이 출력됨

r1, r2, r3 = reverse('a', 'b', 'c')  
print(r1,r2,r3)
print(r1); print(r2); print(r3)      # 튜플의 요소 개수만큼 나누어서 리턴값을 개별적으로 받을 수 있음

(3, 2, 1)
c b a
c
b
a


In [27]:
# 현재 작업 폴더
%pwd

'C:\\hmkd1\\m1_programming'

In [55]:
# ls
!dir

 D 드라이브의 볼륨: 새 볼륨
 볼륨 일련 번호: 02B1-19A2

 D:\kdt_231026\workspace\m1_programming 디렉터리

2023-11-01  오후 02:22    <DIR>          .
2023-11-01  오후 02:22    <DIR>          ..
2023-10-31  오후 12:53    <DIR>          .ipynb_checkpoints
2023-11-01  오후 02:22    <DIR>          ex
2023-10-30  오후 02:02            59,057 python_1_자료형.ipynb
2023-11-01  오전 09:45           102,031 python_2_제어문.ipynb
2023-11-01  오후 02:21            82,542 python_3_함수.ipynb
2023-10-31  오후 12:53           433,932 python_4_함수응용.ipynb
               4개 파일             677,562 바이트
               4개 디렉터리  435,521,261,568 바이트 남음


In [51]:
# 파일 만들기
!echo. > test.txt

In [53]:
# 파일 삭제
!del test.txt

#### 파일 입출력

In [59]:
# 폴더 경로 이동
import os
os.chdir('D:\\kdt_231026\\workspace\\m1_programming')

In [60]:
%pwd

'D:\\kdt_231026\\workspace\\m1_programming'

In [61]:
# 파일 읽고 쓰기
f = open('test1.txt','w')
f.close()

In [62]:
ls

 D 드라이브의 볼륨: 새 볼륨
 볼륨 일련 번호: 02B1-19A2

 D:\kdt_231026\workspace\m1_programming 디렉터리

2023-11-01  오후 02:25    <DIR>          .
2023-11-01  오후 02:25    <DIR>          ..
2023-10-31  오후 12:53    <DIR>          .ipynb_checkpoints
2023-11-01  오후 02:22    <DIR>          ex
2023-10-30  오후 02:02            59,057 python_1_자료형.ipynb
2023-11-01  오전 09:45           102,031 python_2_제어문.ipynb
2023-11-01  오후 02:25            82,635 python_3_함수.ipynb
2023-10-31  오후 12:53           433,932 python_4_함수응용.ipynb
2023-11-01  오후 02:25                 0 test1.txt
               5개 파일             677,655 바이트
               4개 디렉터리  435,521,261,568 바이트 남음


In [63]:
f = open('test1.txt','w')
for i in range(1,11):
    data = '%d번째 줄입니다.\n' %i
    f.write(data)
f.close()

In [34]:
f = open('test1.txt','r')
line = f.readline()
print(line)
f.close()

1번째 줄입니다.



In [35]:
f = open('test1.txt', 'r')
for i in range(10):
    line = f.readline()
    print(line)
f.close()

1번째 줄입니다.

2번째 줄입니다.

3번째 줄입니다.

4번째 줄입니다.

5번째 줄입니다.

6번째 줄입니다.

7번째 줄입니다.

8번째 줄입니다.

9번째 줄입니다.

10번째 줄입니다.



In [65]:
f = open('test1.txt','r')
for i in range(10):
    line = f.readline()
    if i == 4:
        print(line)
f.close()

5번째 줄입니다.



In [17]:
f = open('test1.txt','r')
while True:
    line = f.readline()
    if not line: break
    print(line)
f.close()

1번째 줄입니다.

2번째 줄입니다.

3번째 줄입니다.

4번째 줄입니다.

5번째 줄입니다.

6번째 줄입니다.

7번째 줄입니다.

8번째 줄입니다.

9번째 줄입니다.

10번째 줄입니다.



In [66]:
# f.read()는 파일 내용 전체를 문자열로 반환
f = open('test1.txt','r')
data = f.read()
print(data)
f.close()

1번째 줄입니다.
2번째 줄입니다.
3번째 줄입니다.
4번째 줄입니다.
5번째 줄입니다.
6번째 줄입니다.
7번째 줄입니다.
8번째 줄입니다.
9번째 줄입니다.
10번째 줄입니다.



In [67]:
f = open('test1.txt', 'a')
for i in range(11,16):
    data = '%d번째 줄입니다.\n' %i
    f.write(data)
f.close()

In [68]:
f = open('test1.txt','r')
data = f.read()
print(data)
f.close()

1번째 줄입니다.
2번째 줄입니다.
3번째 줄입니다.
4번째 줄입니다.
5번째 줄입니다.
6번째 줄입니다.
7번째 줄입니다.
8번째 줄입니다.
9번째 줄입니다.
10번째 줄입니다.
11번째 줄입니다.
12번째 줄입니다.
13번째 줄입니다.
14번째 줄입니다.
15번째 줄입니다.



In [69]:
# with문과 같이 사용
with open('foo.txt','w') as f:
    f.write('파이썬을 배웁시다.')

In [70]:
with open('foo.txt','r') as f:
    data = f.read()
    print(data)

파이썬을 배웁시다.


In [71]:
with open('test2.txt','w') as f:
      for i in range(1,11):
          data = f'{i}번째 줄입니다.\n'
          f.write(data)          

In [72]:
with open('test2.txt','a') as f:
      for i in range(11,16):
          data = f'{i}번째 줄입니다.\n'
          f.write(data) 

In [73]:
with open('test2.txt','r') as f:
    data = f.read()
    print(data)

1번째 줄입니다.
2번째 줄입니다.
3번째 줄입니다.
4번째 줄입니다.
5번째 줄입니다.
6번째 줄입니다.
7번째 줄입니다.
8번째 줄입니다.
9번째 줄입니다.
10번째 줄입니다.
11번째 줄입니다.
12번째 줄입니다.
13번째 줄입니다.
14번째 줄입니다.
15번째 줄입니다.



In [15]:
with open('want1.txt', 'w', encoding='utf-8') as f:
    f.write('가나다라마바사아자차카타파하')


with open('want1.txt', 'r', encoding='utf-8') as z:
    data = z.read()
    print(data)

가나다라마바사아자차카타파하


In [ ]:
# 과제5_1101 : 원하는 데이터를 생성해서 저장하고 다시 읽어오세요.

#### 파이썬 패키지,  모듈, 함수, 클래스
- 코드 작성 시 이미 만들어져 있는 함수들을 활용하면 보다 효율적이고 빠르게 개발 가능
- 이미 만들어져 있고 안정성이 검증된 함수들을 성격에 맞게 하나의 파이썬 파일에 묶어 만들어 놓을 것을 모듈이라 함
- 외부 모듈에 있는 함수들을 활용하려면 이 모듈을 먼저 가져와서 사용해야 하는데 이것을 모듈을 임포트한다고 한다.
- 파이썬 모듈이 하나의 파이썬 파일이라면 패키지는 디렉토리로 볼 수 있다.

패키지
1. 라이브러리라고도 부름
2. 특정 기능과 관련된 여러 모듈을 한 그룹으로 묶은 것
3. 패키지 안에 서브패키지가 있을 수도 있다.
4. import 패키지/모듈, from 패키지/모듈 import 서브패키지/모듈 
  
   
모듈
1. 여러 기능들이 뭉쳐진 하나의. py파일
2. 함수, 클래스, 변수 등 포함
3. import 모듈

함수
1. 하나의 기능을 가진 코드의 집합
2. 함수를 사용한다 = 함수를 호출한다
3. 함수 사용 형태: 함수( )
   - 필요한 경우 () 안에 매개변수 또는 옵션
4. 종류
   - 내장 함수(built in function)
   - 외장 함수: import 해주어야 함
   
클래스   
- 파이썬은 객체 지향 언어로클래스를 기반으로 객체를 만들어 사용한다.
- 한 파이썬 파일(. py) 내에서 "class class명():"을 통해 클래스 선언


In [74]:
# time 모듈
import time

print('5초간 프로그램을 정지합니다.')
time.sleep(5)
print('5초가 지나갔습니다.')

5초간 프로그램을 정지합니다.
5초가 지나갔습니다.


In [79]:
ls

 D 드라이브의 볼륨: 새 볼륨
 볼륨 일련 번호: 02B1-19A2

 D:\kdt_231026\workspace\m1_programming 디렉터리

2023-11-01  오후 03:16    <DIR>          .
2023-11-01  오후 03:16    <DIR>          ..
2023-10-31  오후 12:53    <DIR>          .ipynb_checkpoints
2023-11-01  오후 02:22    <DIR>          ex
2023-11-01  오후 02:43                18 foo.txt
2023-11-01  오후 03:16                84 mymodule.py
2023-10-30  오후 02:02            59,057 python_1_자료형.ipynb
2023-11-01  오전 09:45           102,031 python_2_제어문.ipynb
2023-11-01  오후 03:15            85,812 python_3_함수.ipynb
2023-10-31  오후 12:53           433,932 python_4_함수응용.ipynb
2023-11-01  오후 02:40               261 test1.txt
2023-11-01  오후 02:44               261 test2.txt
               8개 파일             681,456 바이트
               4개 디렉터리  435,521,261,568 바이트 남음


In [78]:
# 사용자 작성 모듈
with open('mymodule.py','w') as f:
    f.write(
'''
def add_txt(t1,t2):
    return t1+':'+t2
def reverse(x,y,z):
    return z,y,x
''')

In [80]:
with open('mymodule.py','r') as f:
    data = f.read()
    print(data)


def add_txt(t1,t2):
    return t1+':'+t2
def reverse(x,y,z):
    return z,y,x



In [81]:
import mymodule

ret1 = mymodule.add_txt('대한민국','1등')
print(ret1)

대한민국:1등


In [82]:
ret2 = mymodule.reverse(1,2,3)
print(ret2)

(3, 2, 1)


In [83]:
# 클래스
class MyClass:
    var = '안녕하세요'
    def sayHello(self):
        print(self.var)
        
obj = MyClass() # MyClass 인스턴스 객체 생성
print(obj.var)
print(obj.sayHello())       

안녕하세요
안녕하세요
None


In [ ]:
# !pip list
# !pip install pandas

# https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html

In [86]:
import pandas as pd

df1 = pd.DataFrame({'a': [1, 2, 3], 'b': [4,5,6], 'c': [7, 8,9]})
df1

a  b  c
0  1  4  7
1  2  5  8
2  3  6  9

In [94]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   a       3 non-null      int64
 1   b       3 non-null      int64
 2   c       3 non-null      int64
dtypes: int64(3)
memory usage: 200.0 bytes


In [89]:
import numpy as np
df2 = pd.DataFrame(np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]]),
                   columns=['a', 'b', 'c'])
df2

a  b  c
0  1  2  3
1  4  5  6
2  7  8  9

In [90]:
data = np.array([(1, 2, 3), (4, 5, 6), (7, 8, 9)])
df3 = pd.DataFrame(data, columns=['a', 'b', 'c'])
df3

a  b  c
0  1  2  3
1  4  5  6
2  7  8  9

내장함수, 외장함수

https://codedrive.tistory.com/91

In [ ]:
# 내장함수
print(abs(-3))
print(all([1,2,3,0])) # x가 모두 참이면 True, 거짓이 하나라도 있으면 False
print(any([1,2,3,0])) # x가 하나라도 참이면 True, 모두 거짓이면 False
print(chr(65)) # 아스키코드

3
False
True
A


In [51]:
print(float(3))
print(int(3.9))
print(max(1,10))
print(min(1,10))
print(pow(10,2)) # 거듭제곱
print(round(2.675))

3.0
3
10
1
100
3


In [1]:
li = [1,2,3]
sum(li)

6

In [100]:
my_list = [1, 2, 3, 4, 5]
print(len(my_list)) 

5


In [95]:
age = 30
print("I am " + str(age) + " years old.") 

I am 30 years old.


In [96]:
my_tuple = (1, 2, 3)
print(list(my_tuple))

[1, 2, 3]


In [98]:
my_list = [3, 1, 4, 2]
print(sorted(my_list))
my_list

[1, 2, 3, 4]


[3, 1, 4, 2]

In [99]:
my_string = "Python"
print(type(my_string))

<class 'str'>


In [3]:
for idx, element in enumerate(['ele1','ele2','ele3']):
    print(idx+1,element)

1 ele1
2 ele2
3 ele3


In [5]:
print(range(10))
list(range(10))

range(0, 10)


[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [101]:
# 몫과 나머지
print(divmod(7,3))

(2, 1)


In [17]:
!echo. > mydata.txt
with open('mydata.txt','w') as f:
    f.write(
'''
a lot of things occur each day
we've got to live with on our way
there's so much misery on the screen
with commercials in between

and the russians won't be so late
uncle sam don't want to wait
no matter if you're black or white
you never ever had a right

and look up to the eternal skies
see warfare even there
what once has been a paradise
now destruction and despair

in other worlds the children die
lacking food ill from a fly
oppressed by troops to tame their land
it's all the same again

if any man is free on earth
then tormentors steal his birth
they slam gun butts in his face
and leave his woman choked in mace

so stand up for our human rights
push back this pest of hate
raise you voice pass on the light
unite! it's not too late

how many tears flow away
to become a sea of fears
how many hearts are torn apart
till another torment starts

but before the world
turns into a sun
all cruelty and violence
on earth will be dead and gone

so stand up for our human rights
push back this pest of hate
raise you voice pass on the light
unite! it's not too late
''')
with open('mydata.txt', 'r') as f:
    data = f.read()
#     print(data)
words_list = data # 데이터 받음
words_list = list(words_list.replace(',', '').replace('.', '').replace('\n', '').replace("'", "").replace('!','').replace(' ','')) # 잡다한거 지우고 리스트화
words_set = set(words_list) # 리스트를 세트화한것을 변수에 저장
words_listed_set = list(words_set) # 세트화한 변수를 리스트화
word_frequency = {word : 0 for word in words_listed_set} # 세트화한 변수를 리스트화한 리스트를 딕셔너리화
for word in words_list: # 글자가 있는 리스트에서 글자하나씩 떼고
    if word in word_frequency: # 딕셔너리안에 있는 글자들도 하나씩 떼서
        word_frequency[word] += 1 # 딕셔너리[글자] = 갯수 - 돌린다.
word_frequency = sorted(word_frequency.items(), key= lambda x: x[1], reverse=True) # 많은 순서대로 나열했습니다. 순서를 가장 많은것부터 오게 첫번째부터
print('=' * 50)
print('가장 많이 나온 단어 5개를 뽑아 봤습니다.')
for i in range(5):
    print(f'[{word_frequency[i][0]}] -> [{word_frequency[i][1]}]회 나타남')

가장 많이 나온 단어 5개를 뽑아 봤습니다.
[e] -> [97]회 나타남
[t] -> [85]회 나타남
[a] -> [78]회 나타남
[o] -> [73]회 나타남
[s] -> [60]회 나타남


In [18]:
read_all = ''
with open('mydata.txt','r') as f:
    read_all = f.read()
    
str_set = set(read_all)
str_dict = {}

for i in str_set:
    str_dict[i] = read_all.count(i)

for k, v in str_dict.items():
    if k == '\n':
        k = '\\n'
        
    if k == ' ':
        print(f'[{k:2}] -> [{v:3}]회 나타남 (해당란은 공백값의 사용횟수를 표시)')
    elif k == '\\n':
        print(f'[{k:2}] -> [{v:3}]회 나타남 (해당란은 줄바꿈의 사용횟수를 표시)')
    else:
        print(f'[{k:2}] -> [{v:3}]회 나타남')

[\n] -> [ 45]회 나타남 (해당란은 줄바꿈의 사용횟수를 표시)
[a ] -> [ 78]회 나타남
[r ] -> [ 54]회 나타남
[h ] -> [ 48]회 나타남
[s ] -> [ 60]회 나타남
[u ] -> [ 27]회 나타남
[i ] -> [ 50]회 나타남
[e ] -> [ 97]회 나타남
[d ] -> [ 24]회 나타남
[b ] -> [ 13]회 나타남
[f ] -> [ 17]회 나타남
[! ] -> [  2]회 나타남
[  ] -> [177]회 나타남 (해당란은 공백값의 사용횟수를 표시)
[k ] -> [  7]회 나타남
[l ] -> [ 34]회 나타남
[n ] -> [ 60]회 나타남
[m ] -> [ 20]회 나타남
[g ] -> [ 11]회 나타남
[o ] -> [ 73]회 나타남
[' ] -> [  8]회 나타남
[w ] -> [ 20]회 나타남
[c ] -> [ 23]회 나타남
[v ] -> [  9]회 나타남
[p ] -> [ 15]회 나타남
[t ] -> [ 85]회 나타남
[y ] -> [ 15]회 나타남


도전과제6_1101 :
텍스트 파일(mydata.txt)을 읽고  파일의 내용에서 나타나는 문자들에 대해 자주 나타나는 문자들을 아래와 같이 순서대로 출력하세요.
- \[ \] -> [177]회 나타남
- [e] -> [97]회 나타남
- [t] -> [85]회 나타남

In [102]:
with open('mydata.txt','w') as f:
    f.write(
'''
a lot of things occur each day
we've got to live with on our way
there's so much misery on the screen
with commercials in between

and the russians won't be so late
uncle sam don't want to wait
no matter if you're black or white
you never ever had a right

and look up to the eternal skies
see warfare even there
what once has been a paradise
now destruction and despair

in other worlds the children die
lacking food ill from a fly
oppressed by troops to tame their land
it's all the same again

if any man is free on earth
then tormentors steal his birth
they slam gun butts in his face
and leave his woman choked in mace

so stand up for our human rights
push back this pest of hate
raise you voice pass on the light
unite! it's not too late

how many tears flow away
to become a sea of fears
how many hearts are torn apart
till another torment starts

but before the world
turns into a sun
all cruelty and violence
on earth will be dead and gone

so stand up for our human rights
push back this pest of hate
raise you voice pass on the light
unite! it's not too late
''')

In [105]:
def getTextFreq(filename):
    with open(filename,'r') as f:
        text = f.read()
        fa = {}
        for c in text:
            if c in fa:
                fa[c] += 1
            else:
                fa[c] = 1
    return fa
ret = getTextFreq('mydata.txt')
print(ret.items(),'\n')
ret = sorted(ret.items(), key=lambda x:x[1],reverse=True)

for c, freq in ret:
    if c == '\n':
        continue
    print(f'[{c}] -> [{freq}]회 나타남')

dict_items([('\n', 45), ('a', 78), (' ', 177), ('l', 34), ('o', 73), ('t', 85), ('f', 17), ('h', 48), ('i', 50), ('n', 60), ('g', 11), ('s', 60), ('c', 23), ('u', 27), ('r', 54), ('e', 97), ('d', 24), ('y', 15), ('w', 20), ("'", 8), ('v', 9), ('m', 20), ('b', 13), ('k', 7), ('p', 15), ('!', 2)]) 

[ ] -> [177]회 나타남
[e] -> [97]회 나타남
[t] -> [85]회 나타남
[a] -> [78]회 나타남
[o] -> [73]회 나타남
[n] -> [60]회 나타남
[s] -> [60]회 나타남
[r] -> [54]회 나타남
[i] -> [50]회 나타남
[h] -> [48]회 나타남
[l] -> [34]회 나타남
[u] -> [27]회 나타남
[d] -> [24]회 나타남
[c] -> [23]회 나타남
[w] -> [20]회 나타남
[m] -> [20]회 나타남
[f] -> [17]회 나타남
[y] -> [15]회 나타남
[p] -> [15]회 나타남
[b] -> [13]회 나타남
[g] -> [11]회 나타남
[v] -> [9]회 나타남
['] -> [8]회 나타남
[k] -> [7]회 나타남
[!] -> [2]회 나타남


In [19]:
sum = lambda a,b:a+b
sum(3,4)

7

In [20]:
def sum(a,b):
    return a+b

print(sum(3,4))

7


In [21]:
myList = [lambda a,b:a+b, lambda a,b:a*b]
print(myList[0](3,4))
print(myList[1](3,4))

7
12


In [22]:
# Q . lambda를 리스트 형태로 작성하여 4칙 연산을 해보세요.

list1 = [lambda a,b:a+b, lambda a,b:a-b, lambda a,b:a*b, lambda a,b:a/b]
for i in range(0,4):
    print(list1[i](11,2))

13
9
22
5.5


In [3]:
fo=[lambda a,b:a+b,lambda a,b:a-b,lambda a,b:a*b,lambda a,b:a/b]
for i in range(0,4):
    print(fo[i](2,5))

7
-3
10
0.4


In [ ]:
# 과제1_1102 : 실행 가능한 숫자를 문자열('5+5') 입력으로 받아 문자열을 실행한 결과값을 출력하세요.

In [58]:
print(eval(input('연산을 입력하세요.: ')))

연산을 입력하세요.: 5*5
25


In [1]:
num = input('5+5 : ')
maths = ['+','-','/','*']
math = {'+':lambda a,b:a+b,'-':lambda a,b:a-b,'/':lambda a,b:a/b,'*':lambda a,b:a*b}

num_list = list(num)
for i in num_list:
    if i in maths:
        num = num.split(i)
        print(math[i](int(num[0]),int(num[1])))

5+5 : 5+5
10


In [23]:
print(len('python'))
print(len([1,2,3]))
print(len([1,2,'a']))

6
3
3


In [28]:
li = [1,2,3,4,5]
   
for i in range(1,len(li)+1):
    print(i,end=' ')
print()
print(len(li))

1 2 3 4 5 
5


In [29]:
print(list(range(5)))
print(list(range(1,6)))
print(list(range(1,10, 2)))
print(list(range(0, -10, -1)))

[0, 1, 2, 3, 4]
[1, 2, 3, 4, 5]
[1, 3, 5, 7, 9]
[0, -1, -2, -3, -4, -5, -6, -7, -8, -9]


In [ ]:
# 순서가 있는 자료형을 입력으로 받아 인덱스 값을 포함한 enumerate 객체를 리턴
for i, name in enumerate(['james','tom','susan']):
    print(i,name)

0 james
1 tom
2 susan


In [ ]:
# 과제2_1102 : enumerate와 f-string을 사용해서 특정 리스트에서 원하는 형태로 출력을 하세요.(인데스와 내용을 믹스하여)

In [2]:
list02 = ['사과','키위','딸기','참외','수박']

print("과일의 종류")
for i, a in enumerate(list02):
    print(f' {i}. {a}')

과일의 종류
 0. 사과
 1. 키위
 2. 딸기
 3. 참외
 4. 수박


In [3]:
list1 = ['김부장', '이차장', '김과장', '이대리', '오사원', '김인턴']

for index, name in enumerate(list1):
    print(f'인사평가 {index+1}번째 해당자는 {name}입니다. 회의실A로 오시기 바랍니다.')

인사평가 1번째 해당자는 김부장입니다. 회의실A로 오시기 바랍니다.
인사평가 2번째 해당자는 이차장입니다. 회의실A로 오시기 바랍니다.
인사평가 3번째 해당자는 김과장입니다. 회의실A로 오시기 바랍니다.
인사평가 4번째 해당자는 이대리입니다. 회의실A로 오시기 바랍니다.
인사평가 5번째 해당자는 오사원입니다. 회의실A로 오시기 바랍니다.
인사평가 6번째 해당자는 김인턴입니다. 회의실A로 오시기 바랍니다.


In [ ]:
# 실행 가능한 문자열 입력으로 받아 문자열을 실행한 결과값을 리턴
print(eval('1+2'))

3


In [30]:
# 외장함수 : 전 세계 파이썬 사용자들이 만든 유용한 프로그램을 모아 놓은 파이썬 라이브러리. 파이썬을 설치할 때 자동으로 설치
# pickle은 객체의 형태를 유지하면서 파일에 저장하고 불러올 수 있게 하는 모듈

import pickle
f = open('test3.txt', 'wb')
data = {1:'python',2:'you need'}
pickle.dump(data,f)
f.close()

In [31]:
import pickle
f = open('test3.txt','rb')
data = pickle.load(f)
print(data)
f.close()

{1: 'python', 2: 'you need'}


In [36]:
import os
os.environ['PATH']

'C:\\ProgramData\\Anaconda3\\envs\\kita;C:\\ProgramData\\Anaconda3\\envs\\kita\\Library\\mingw-w64\\bin;C:\\ProgramData\\Anaconda3\\envs\\kita\\Library\\usr\\bin;C:\\ProgramData\\Anaconda3\\envs\\kita\\Library\\bin;C:\\ProgramData\\Anaconda3\\envs\\kita\\Scripts;C:\\Program Files\\Java\\jdk-1.8\\bin;C:\\Windows\\system32;C:\\Windows;C:\\Windows\\System32\\Wbem;C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\;C:\\Windows\\System32\\OpenSSH\\;C:\\ProgramData\\Anaconda3\\Scripts;C:\\ProgramData\\Anaconda3\\Library\\bin;C:\\ProgramData\\Anaconda3\\Library\\mingw-w64\\bin;C:\\ProgramData\\Anaconda3;C:\\Users\\Administrator\\AppData\\Local\\Microsoft\\WindowsApps;C:\\Program Files\\Bandizip\\;C:\\Users\\Administrator\\AppData\\Local\\Programs\\Microsoft VS Code\\bin'

In [42]:
%pwd

'D:\\kdt_231026\\workspace\\m1_programming'

In [39]:
# 폴더변경
os.chdir('D:\\kdt_231026\\workspace\\m1_programming\\ex')

In [41]:
os.chdir('D:\\kdt_231026\\workspace\\m1_programming')

In [43]:
os.getcwd()

'D:\\kdt_231026\\workspace\\m1_programming'

In [49]:
!dir

 D 드라이브의 볼륨: 새 볼륨
 볼륨 일련 번호: 02B1-19A2

 D:\kdt_231026\workspace\m1_programming 디렉터리

2023-11-02  오전 11:06    <DIR>          .
2023-11-02  오전 11:06    <DIR>          ..
2023-10-31  오후 12:53    <DIR>          .ipynb_checkpoints
2023-11-01  오후 02:22    <DIR>          ex
2023-11-01  오후 02:43                18 foo.txt
2023-11-02  오전 10:12             1,117 mydata.txt
2023-11-01  오후 03:16                84 mymodule.py
2023-10-30  오후 02:02            59,057 python_1_자료형.ipynb
2023-11-01  오전 09:45           102,031 python_2_제어문.ipynb
2023-11-02  오전 11:05           105,285 python_3_함수.ipynb
2023-10-31  오후 12:53           433,932 python_4_함수응용.ipynb
2023-11-01  오후 02:40               261 test1.txt
2023-11-01  오후 02:44               261 test2.txt
2023-11-02  오전 10:49                40 test3.txt
2023-11-02  오전 10:08                42 want1.txt
2023-11-01  오후 03:19    <DIR>          __pycache__
              11개 파일             702,128 바이트
               5개 디렉터리  435,586,596,864 바이트 남음


In [48]:
import os
os.rename('test_new.txt','test1.txt')

In [50]:
# glob() 함수는 인자로 받은 패턴과 이름이 일치하는 모든 파일과 디렉토리의 리스트를 반환
import glob
glob.glob('D:\\kdt_231026\\workspace\\m1_programming/*.txt')

['D:\\kdt_231026\\workspace\\m1_programming\\foo.txt',
 'D:\\kdt_231026\\workspace\\m1_programming\\mydata.txt',
 'D:\\kdt_231026\\workspace\\m1_programming\\test1.txt',
 'D:\\kdt_231026\\workspace\\m1_programming\\test2.txt',
 'D:\\kdt_231026\\workspace\\m1_programming\\test3.txt',
 'D:\\kdt_231026\\workspace\\m1_programming\\want1.txt']

In [ ]:
# 과제3_1102 : 내장함수를 이용해서 list = [0,1,2,3,4,5,6]에서 2를 삭제 후 출력하세요.(3가지 방법)

In [4]:
#1번째 방법
list1 = [0,1,2,3,4,5,6]
del list1[2]
print(list1)

#2번째 방법
list1 = [0,1,2,3,4,5,6]
list1.remove(2)
print(list1)

#3번째 방법
list1 = [0,1,2,3,4,5,6]
list1.pop(2)
print(list1)

[0, 1, 3, 4, 5, 6]
[0, 1, 3, 4, 5, 6]
[0, 1, 3, 4, 5, 6]


In [11]:

list=[0,1,2,3,4,5,6]
# del list[2] # 인덱스
list.remove(2) # 값
# list.pop(2)
print(list)

[0, 1, 3, 4, 5, 6]


In [14]:
list=[0,1,2,3,4,5,6]
print(list.pop(2))
list

2


[0, 1, 3, 4, 5, 6]

In [ ]:
# 과제4_1102 :  내장함수를 이용해서 list = [0,1,2,3,4,5,6]에서 모든 원소를 삭제 후 출력하세요.

In [5]:
my_list = [0, 1, 2, 3, 4, 5, 6]
del my_list[:]
print(my_list)

[]


In [56]:

list = [0,1,2,3,4,5,6]
print(list)
del list[:]
print(list)

[0, 1, 2, 3, 4, 5, 6]
[]


In [ ]:
# 과제5_1102 :  람다함수와 적합한 내장함수를 사용하여 아래와 같이 출력하세요
[1,2,3,4,5] -> [1,4,9,16,25]
[1,2,3,4,5] -> [1,2]

In [7]:
lst = [1, 2, 3, 4, 5]

# 람다 함수를 사용하여 리스트의 각 요소를 제곱합니다.
squared_lst = list(map(lambda x: x**2, lst))

# 새로운 리스트를 생성하여, 원래 리스트의 첫 번째와 두 번째 요소만 포함합니다.
# new_lst = lst[:2]
del(lst[2:])

# 결과를 출력합니다.
print(squared_lst)
print(lst)

[1, 4, 9, 16, 25]
[1, 2]


In [35]:
list=[1,2,3,4,5]
list_n=[]
x=lambda a:a**2  

for i in list:          # [1,2,3,4,5] -> [1,4,9,16,25]
    list_n.append(x(i))
   
del(list[2:])           # [1,2,3,4,5] -> [1,2]

print(list_n)         #함수호출
print(list)

[1, 4, 9, 16, 25]
[1, 2]


In [1]:
x=lambda a:a**2
lst = [x(i) for i in range(1,6)]
print(lst,'\n')

lst2 = [1,2,3,4,5]
filtered_list = filter(lambda a: (a<3), lst2)
print(list(filtered_list))

[1, 4, 9, 16, 25] 

[1, 2]


In [ ]:
# 과제6_1102  [1,-3,2,0,-5,6] 에서 양수만 필터링해서 출력하세요(내장함수, 사용자함수 각각 이용)

In [12]:
# 내장함수
list2 = [1,-3,2,0,-5,6] 
result = filter(lambda x : x > 0, list2) 
print(list(result))

# 사용자 함수
def even(arr):
    result=[]
    for i in arr:
        if i>0:
            result.append(i)
    return result
print(even([1,-3,2,0,-5,6]))
print(even([1,-3,2,-5,-6]))

# 간편한 for문
sample6 = [1,-3,2,0,-5,6]
ans6_2 = [i for i in sample6 if i > 0]
print(ans6_2)

[1, 2, 6]
[1, 2, 6]
[1, 2]
[1, 2, 6]


In [13]:
list1 = [1,2,3]
list2 = [4,5,6]
list3 = [7,8,9]
list4 = []
list5 = []
list6 = []
t1 = ('abc','def')

# 1번
for i in range(3):
    new = list1[i], list2[i]
    list4.insert(i, new)
    i += 1

print(f'{list1},{list2} → {list4}')

# 2번
for i in range(3):
    new1 = list1[i], list2[i], list3[i]
    list5.insert(i, new1)
    i += 1

print(f'{list1},{list2},{list3} → {list5}')

# 3번
trn1 = list(t1[0])
trn2 = list(t1[1])

for i in range(3):
    new2 = trn1[i], trn2[i]
    list6.insert(i, new2)
    i += 1

print(f'{t1} → {list6}')

[1, 2, 3],[4, 5, 6] → [(1, 4), (2, 5), (3, 6)]
[1, 2, 3],[4, 5, 6],[7, 8, 9] → [(1, 4, 7), (2, 5, 8), (3, 6, 9)]
('abc', 'def') → [('a', 'd'), ('b', 'e'), ('c', 'f')]


In [ ]:
# 과제7_1102 : 내장함수를 이용하여 동일한 개수로 이루어진 자료형을 화살표 좌측에서 우측으로 변환하세요
[1,2,3],[4,5,6] -> [(1,4),(2,5),(3,6)]
[1,2,3],[4,5,6] [7,8,9] -> [(1,4,7),(2,5,8),(3,6,9)]
('abc','def') -> [('a','d'),('b','e'),('c','f')]

In [15]:
list5 = [1,2,3]
list6 = [4,5,6]
list7 = [7,8,9]
tuple8 = ('abc','def')
print([i for i in zip(list5, list6)])
print([i for i in zip(list5, list6, list7)])
print([i for i in zip(tuple8[0],tuple8[1])])

[(1, 4), (2, 5), (3, 6)]
[(1, 4, 7), (2, 5, 8), (3, 6, 9)]
[('a', 'd'), ('b', 'e'), ('c', 'f')]


In [7]:
a = [1, 2, 3]
b = [4, 5, 6]
c = [7, 8, 9]

ab = list(zip([1,2,3],[4,5,6]))
print(ab)
                        
abc = list(zip([1,2,3],[4,5,6], [7,8,9]))
print(abc)

alp_1 = list(zip('abc', 'def'))
print(alp_1)

[(1, 4), (2, 5), (3, 6)]
[(1, 4, 7), (2, 5, 8), (3, 6, 9)]
[('a', 'd'), ('b', 'e'), ('c', 'f')]


In [6]:
lista=[[1,2,3],[4,5,6],[7,8,9]]
listb=sum(lista,[])
lista_n=[]

for k in range(len(lista[0])):
    tup1=tuple(listb[k::3])
    lista_n.append(tup1)

print(lista_n)

[(1, 4, 7), (2, 5, 8), (3, 6, 9)]


In [ ]:
# 과제8_1102 : 입력으로 들어오는 모든 수의 평균 값을 계산해 주는 함수를 작성한 후 1,2,3을 입력한 결과를 출력하세요.

In [16]:
def cal_average(*args):
    return sum(args) / len(args)

print(cal_average(1,2,3))

2.0


In [23]:
def str_average(str):
    mun_str = str.replace(',','+')
    print(eval(mun_str)/(mun_str.count('+')+1))
    
mun = input('평균값을 구할 숫자를 모두 입력해주세요(콤마로 구분): ')
str_average(mun)

평균값을 구할 숫자를 모두 입력해주세요(콤마로 구분): 1,2,3
2.0


In [24]:
def cal_list_avg(avgList):
    y = avgList.split(',')
    x = [int(i) for i in y]
    return sum(x) / len(x)

ans = input('평균값을 구할 숫자(콤마로 구분하여 입력): ')
print(cal_list_avg(ans))

평균값을 구할 숫자(콤마로 구분하여 입력): 1,2,3
2.0


In [22]:
def cal_avg(mun):
    print(f'합: {eval(mun)}')
    print(f"평균: {eval(mun)/len(mun.split('+')):0.2f}")
# cal_avg('1+2+3')

mun = input('평균값을 구할 숫자를 모두 입력해주세요(콤마로 구분): ')

합: 6
평균: 2.00


In [ ]:
# 과제9_1102 : 동일한 사용자 함수를 만들어서 1~100의 합과 200~300의 합을 구하세요.

In [25]:
def myCal_sum(sInt, eInt):
    sum = 0
    
    for i in range(sInt, eInt+1):
        sum += i
    
    return print(f"{sInt}부터 {eInt}까지의 합은 {sum}입니다.")

myCal_sum(1, 100)
myCal_sum(200, 300)

1부터 100까지의 합은 5050입니다.
200부터 300까지의 합은 25250입니다.


In [26]:
def add(start, end):
    total = sum(range(start, end+1))
    return total

sum100 = add(0, 100)
sum200 = add(200, 300)

print(sum100, sum200)

5050 25250


In [51]:
def sum(first, end):
    result=0
    for i in range(first,end+1):
        result+=i
    return result

print(sum(100,200))   #n~m 사이의 합 구하기
print(sum(200,300))

15150
25250


In [27]:
import numpy as np

np.arange(200,301)

array([200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212,
       213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225,
       226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238,
       239, 240, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 251,
       252, 253, 254, 255, 256, 257, 258, 259, 260, 261, 262, 263, 264,
       265, 266, 267, 268, 269, 270, 271, 272, 273, 274, 275, 276, 277,
       278, 279, 280, 281, 282, 283, 284, 285, 286, 287, 288, 289, 290,
       291, 292, 293, 294, 295, 296, 297, 298, 299, 300])

In [28]:
import numpy as np
def sum(*args):
    sum=0
    for i in args:
        sum += i
    return sum

sum(200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212,
       213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225,
       226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238,
       239, 240, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 251,
       252, 253, 254, 255, 256, 257, 258, 259, 260, 261, 262, 263, 264,
       265, 266, 267, 268, 269, 270, 271, 272, 273, 274, 275, 276, 277,
       278, 279, 280, 281, 282, 283, 284, 285, 286, 287, 288, 289, 290,
       291, 292, 293, 294, 295, 296, 297, 298, 299, 300)

25250

#### sort, sorted 모두 key, reverse 매개변수를 갖고 있다.
- reverse : 기본값은 reverse=False(오름차순), reverse=True를 매개변수로 입력하면 내림차순으로 정렬
- key : 정렬을 목적으로 하는 함수를 값으로 넣는다. lambda를 이용할 수 있고 key 값을 기준으로 정렬되고 기본값은 오름차순

In [22]:
# strings를 다양한 문자가 포함된 순서로 정렬하세요.
strings = ['a', 'ca', 'bar', 'aaaa', 'ababa']
strings.sort(key = lambda x:len(set(x)))
strings[::-1]

['bar', 'ababa', 'ca', 'aaaa', 'a']

In [30]:
set('ababa')

{'a', 'b'}

In [12]:
# Q. 튜플 형태인 strings를 다양한 문자가 포함된 순서로 정렬하세요,
strings = ('a', 'ca', 'bar', 'aaaa', 'ababa')
s_list=list(strings)
s_list.sort(key=lambda x:len(set(x)))     #기본 정렬(오름차순)
# 'tuple' object has no attribute 'sort'
print(s_list[::-1])     #거꾸로 출력(내림차순으로 나오게)

['bar', 'ababa', 'ca', 'aaaa', 'a']


In [52]:
# url에서 호스트 도메인을 추출하여 출력하세요.
url = 'http://news.naver.com/main/read.nhn?mode=LSD&mid=shm&sid1=105&oid=028&aid=0002334601'

x = url.split('/')
print(x)
x[2]

['http:', '', 'news.naver.com', 'main', 'read.nhn?mode=LSD&mid=shm&sid1=105&oid=028&aid=0002334601']


'news.naver.com'

In [ ]:
# 과제10_1102 : url에서 쿼리문을 추출하여 출력하세요.
# url에서 '?'뒤에 표시되는 문자열을 쿼리 문자열이라고 하며 변수=값 사이 &로 구분되어 나열된다.
url = 'http://news.naver.com/main/read.nhn?mode=LSD&mid=shm&sid1=105&oid=028&aid=0002334601'

In [1]:
url = 'http://news.naver.com/main/read.nhn?mode=LSD&mid=shm&sid1=105&oid=028&aid=0002334601'

tmp1 = url.split('/')
tmp2 = tmp1[-1].split('?')
result = tmp2[-1].split('&')

print(result)
print("")

for i in result:
    print(i)

['mode=LSD', 'mid=shm', 'sid1=105', 'oid=028', 'aid=0002334601']

mode=LSD
mid=shm
sid1=105
oid=028
aid=0002334601


In [53]:
do=url.find('?')
print(do)
print(url[do])
print(url[do+1:])   #쿼리문 전체

qr=[]
qr=url[do+1:].split('&')   #각 '변수=값' 분리
print(qr)

35
?
mode=LSD&mid=shm&sid1=105&oid=028&aid=0002334601
['mode=LSD', 'mid=shm', 'sid1=105', 'oid=028', 'aid=0002334601']


In [54]:
url = 'http://news.naver.com/main/read.nhn?mode=LSD&mid=shm&sid1=105&oid=028&aid=0002334601'
tmp = url.split('?')
print(tmp)
queries = tmp[1].split('&')
for query in queries:
    print(query)

['http://news.naver.com/main/read.nhn', 'mode=LSD&mid=shm&sid1=105&oid=028&aid=0002334601']
mode=LSD
mid=shm
sid1=105
oid=028
aid=0002334601


In [31]:
# 문자열이 문자인지 검사
txt1 = 'A'
txt2 = '안녕'
txt3 = 'Warcraft Three'
txt4 = '3PO'
ret1 = txt1.isalpha()
ret2 = txt2.isalpha()
ret3 = txt3.isalpha()
ret4 = txt4.isalpha()
print(ret1)        # True가 출력됨
print(ret2)        # True가 출력됨
print(ret3)        # False가 출력됨
print(ret4)        # False가 출력됨

True
True
False
False


In [32]:
# 문자열이 숫자인지 검사

txt1 = '010-1234-5678'
txt2 = 'R2D2'
txt3 = '1212'

ret1 = txt1.isdigit()
ret2 = txt2.isdigit()
ret3 = txt3.isdigit()
print(ret1)      # False가 출력됨
print(ret2)      # False가 출력됨
print(ret3)      # True가 출력됨

False
False
True


In [33]:
# 문자열이 문자 또는 숫자인지 검사 : 문자열이 영어, 한글 혹은 숫자로 되어있으면 참 리턴, 아니면 거짓 리턴
txt1 = '안녕하세요?'
txt2 = '1. Title-제목을 넣으세요'
txt3 = '3피오R2D2'

ret1 = txt1.isalnum()
ret2 = txt2.isalnum ()
ret3 = txt3.isalnum ()
print(ret1)      # False가 출력됨
print(ret2)      # False가 출력됨
print(ret3)      # True가 출력됨

False
False
True


In [37]:
txt = 'A lot of Things occur each day.'
ret1 = txt.upper()
ret2 = txt.lower()
ret3 = txt.title()
ret4 = txt.capitalize() # 문자열의 첫번째 문자를 대문자, 나머지는 소문자로 변환
print(ret1)
print(ret2)
print(ret3)
print(ret4)

A LOT OF THINGS OCCUR EACH DAY.
a lot of things occur each day.
A Lot Of Things Occur Each Day.
A lot of things occur each day.
